quickwit-search/src/retry/mod.rs

```diff
@@ -0,0 +1,181 @@
+// Copyright (C) 2021 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+pub mod search;
+pub mod search_stream;
+
+use std::collections::HashSet;
+use std::sync::Arc;
+
+use itertools::Itertools;
+use quickwit_proto::{FetchDocsRequest, LeafSearchRequest, LeafSearchStreamRequest};
+
+use crate::client_pool::Job;
+use crate::{ClientPool, SearchClientPool, SearchServiceClient};
+
+/// A retry policy to evaluate if a request should be retried.
+pub trait RetryPolicy<Request, Response, Error>: Sized
```

Comment:
It was not clear when I read the code, but what happens on how to handle partial success with non-retryable errors or failed retried errors?

(Typically one split is corrupted)

Do we return nothing?
Do we return Ok(...) with whatever data we had and some information in the response to tell what split is missing?
Do we returen Ok(...) silently? 

Maybe the RetryPolicy trait could be a decent place to describe that spec.

Reply:
Good point. I will describe it more accurately.
I described the behaviour for the `LeafSearchRequest` [here](https://github.com/quickwit-inc/quickwit/pull/530/files#diff-721044fceec236ac1242484cf70dddddf96f672c0c7070421d2d48423799a69aR25-R27) but it's rather simple right now.

Basically we want to make a retry either on error or retryable split errors. 
On improvement that can make the code better is just to keep one method `retry_request` that return None if there is no retry and the retry request if we have retryable errors.

Reply:
I saw that comment, but I don't know what the end user sees if one split fails even after retry?

Reply:
Currently we have no HTTP streaming and as soon as we have an error, we stop and return an HTTP error.

With HTTP streaming, we will stream the result and on the first error do something so that the client will understand that something went wrong, I still need to look at that to see what the proper way to stop the stream response.

Reply:
In the `root_search` or `root_search_stream` functions, you have something like this:

```
    let result: Vec<Result> =
        futures::stream::iter(jobs.into_iter())
            .map(|(client, client_jobs)| {
                let leaf_request = jobs_to_leaf_request(client_jobs)
                cluster_client.leaf_search((leaf_request, client))
            })
            .buffer_unordered(MAX_CONCURRENT_LEAF_TASKS)
            .try_collect()
            .await?;
```

Reply:
The rationale is: once some errors reach the root level even after retry, just return propagate the error to the end user.